examples/vision/simsiam.py

```diff
@@ -0,0 +1,434 @@
+        with tf.GradientTape() as tape:
+            z1, z2 = self.encoder(ds_one), self.encoder(ds_two)
+            p1, p2 = self.predictor(z1), self.predictor(z2)
+            # Note that here we are enforcing the network to match
+            # the representations of two differently augmented batches
+            # of data.
+            loss = compute_loss(p1, z2) / 2 + compute_loss(p2, z1) / 2
+
+        # Compute gradients and update the parameters.
+        learnable_params = (
+            self.encoder.trainable_variables + self.predictor.trainable_variables
+        )
+        gradients = tape.gradient(loss, learnable_params)
+        self.optimizer.apply_gradients(zip(gradients, learnable_params))
+
+        # Monitor loss.
+        self.loss_tracker.update_state(loss)
+        return {"loss": self.loss_tracker.result()}
+
+
+"""
+## Pre-training our networks
+
+In the interest of this example, we will train the model for only 5 epochs.
+"""
+
+# Create a cosine decay learning scheduler.
+num_training_samples = len(x_train)
+steps = EPOCHS * (num_training_samples // BATCH_SIZE)
+lr_decayed_fn = tf.keras.experimental.CosineDecay(
+    initial_learning_rate=0.03, decay_steps=steps
+)
+
+# Create an early stopping callback.
+early_stopping = tf.keras.callbacks.EarlyStopping(
+    monitor="loss", patience=5, restore_best_weights=True
+)
+
+# Start training.
+simsiam = SimSiam(get_encoder(), get_predictor())
+simsiam.compile(optimizer=tf.keras.optimizers.SGD(lr_decayed_fn, momentum=0.6))
+history = simsiam.fit(ssl_ds, epochs=EPOCHS, callbacks=[early_stopping])
+
+# Visualize the training progress of the model.
+plt.plot(history.history["loss"])
+plt.grid()
+plt.title("Negative Cosine Similairty")
+plt.show()
+
+"""
+If your solution gets very close to -1 (minimum value of our loss) very quickly with a
+different dataset and a different backbone architecture that is likely because of
+*representation collapse*. It is a phenomenon where the encoder yields similar output for
+all the images. In that case additional hyperparameter tuning is required especially in
+the following areas:
+* Strength of the color distortions and their probabilities. 
+* Learning rate and its schedule. 
+* Architecture of both the backbone and their projection head. 
+"""
+
+"""
+## Evaluating our SSL method
+
+The most popularly used method to evaluate a SSL method in computer vision (or any other
+pre-training method as such) is to learn a linear classifier on the frozen features of
+the trained backbone model (in this case it is ResNet20) and evaluate the classifier on
+unseen images. Other methods include
+[fine-tuning](https://keras.io/guides/transfer_learning/) on the source dataset or even a
+target dataset with 5% or 10% labels present. Practically, we can use the backbone model
+for any downstream task such as semantic segmentation, object detection, and so on where
+the backbone models are usually pre-trained with *pure supervised learning*. 
```

Comment:
Suggestion: some refactoring

```suggestion
The most popular method for evaluating an SSL method in computer vision (or any
other pre-training method as such) is to learn a linear classifier on the frozen
features of the trained backbone model (in our case, it's ResNet-20), and then
evaluate the classifier on unseen images. Other methods include
[fine-tuning](https://keras.io/guides/transfer_learning/) on the source dataset
or even a target dataset with 5% or 10% labels present. We can use the backbone
model for any downstream task, such as semantic segmentation, object detection,
and so on, where the backbone models are usually pre-trained with *pure
supervised learning*. 
```